src/ServiceControl.Persistence.InMemory/UnitOfWork/InMemoryIngestionUnitOfWork.cs

```diff
@@ -0,0 +1,37 @@
+﻿namespace ServiceControl.Persistence.SqlServer
+{
+    using System.Collections.Concurrent;
+    using System.Threading.Tasks;
+    using Monitoring;
+    using Operations;
+
+    class InMemoryIngestionUnitOfWork : IIngestionUnitOfWork
+    {
+        ConcurrentBag<KnownEndpoint> knownEndpoints = new ConcurrentBag<KnownEndpoint>();
+        IMonitoringDataStore dataStore;
+
+        public InMemoryIngestionUnitOfWork(IMonitoringDataStore dataStore)
+        {
+            this.dataStore = dataStore;
+            Monitoring = new InMemoryMonitoringIngestionUnitOfWork(this);
+        }
+
+        public IMonitoringIngestionUnitOfWork Monitoring { get; }
+        public IRecoverabilityIngestionUnitOfWork Recoverability { get; }
+
+        public async Task Complete()
+        {
+            foreach (var endpoint in knownEndpoints)
+            {
+                await dataStore.CreateIfNotExists(endpoint.EndpointDetails)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        internal void AddEndpoint(KnownEndpoint knownEndpoint) => knownEndpoints.Add(knownEndpoint);
+
+        public void Dispose()
```

Comment:
I would naturally seal things because otherwise this class would trigger an analyzer warning (if we had them in place ;)) since it doesn't implement the disposable pattern. Alternative make the disposable pattern opt-in and not part of the interface

Reply:
We moved the disposable implementation into a base class and had that implement the full disposing pattern. That allows the implementations to only override it if they have something to dispose.